core-java-modules/core-java-arrays-operations-basic-2/src/test/java/com/baeldung/array/allelementsareequal/AllElementsAreEqualInArrayUnitTest.java

```diff
@@ -0,0 +1,144 @@
+    private final static String[] ARRAY_ALL_NULL = { null, null, null, null };
+    private final static String[] ARRAY_SINGLE_EL = { "java" };
+
+    // all equal = false arrays:
+    private final static String[] ARRAY_NOT_EQ = { "java", "kotlin", "java", "java" };
+    private final static String[] ARRAY_EMPTY = {};
+    private final static String[] ARRAY_NULL = null;
+
+    //int[] arrays
+    private final static int[] INT_ARRAY_ALL_EQ = { 7, 7, 7, 7 };
+    private final static int[] INT_ARRAY_NOT_EQ = { 7, 7, 7, 42 };
+    private final static int[] INT_ARRAY_SINGLE_EL = { 42 };
+
+    public <T> boolean isAllEqual(T[] array) {
+        if (array == null || array.length == 0) {
+            return false;
+        }
+        for (int i = 1; i < array.length; i++) {
+            if (!Objects.equals(array[0], array[i])) {
+                return false;
+            }
+        }
+        return true;
+    }
+
+    @Test
+    void whenUsingArraysAreEqual_thenCorrect() {
+        assertTrue(isAllEqual(ARRAY_ALL_EQ));
+        assertTrue(isAllEqual(ARRAY_ALL_NULL));
+        assertTrue(isAllEqual(ARRAY_SINGLE_EL));
+
+        assertFalse(isAllEqual(ARRAY_NOT_EQ));
+        assertFalse(isAllEqual(ARRAY_EMPTY));
+        assertFalse(isAllEqual(ARRAY_NULL));
+    }
+
+    public <T> boolean isAllEqualByDistinct(T[] array) {
+        if (array == null || array.length == 0) {
+            return false;
+        }
+        return Arrays.stream(array)
+            .distinct()
+            .count() == 1;
+    }
+
+    @Test
+    void whenUsingArraysAreEqualByDistinct_thenCorrect() {
+        assertTrue(isAllEqualByDistinct(ARRAY_ALL_EQ));
+        assertTrue(isAllEqualByDistinct(ARRAY_ALL_NULL));
+        assertTrue(isAllEqualByDistinct(ARRAY_SINGLE_EL));
+
+        assertFalse(isAllEqualByDistinct(ARRAY_NOT_EQ));
+        assertFalse(isAllEqualByDistinct(ARRAY_EMPTY));
+        assertFalse(isAllEqualByDistinct(ARRAY_NULL));
+    }
+
+    public <T> boolean isAllEqualByAllMatch(T[] array) {
+        if (array == null || array.length == 0) {
+            return false;
+        }
+        return Arrays.stream(array)
+            .allMatch(element -> Objects.equals(array[0], element));
+    }
+
+    @Test
+    void whenUsingArraysAreEqualByAllMatch_thenCorrect() {
+        assertTrue(isAllEqualByAllMatch(ARRAY_ALL_EQ));
+        assertTrue(isAllEqualByAllMatch(ARRAY_ALL_NULL));
+        assertTrue(isAllEqualByAllMatch(ARRAY_SINGLE_EL));
+
+        assertFalse(isAllEqualByAllMatch(ARRAY_NOT_EQ));
+        assertFalse(isAllEqualByAllMatch(ARRAY_EMPTY));
+        assertFalse(isAllEqualByAllMatch(ARRAY_NULL));
+    }
+
+    // primitive array approaches below (int[] as example)
+
+    public boolean isAllEqual(int[] array) {
+        if (array == null || array.length == 0) {
+            return false;
+        }
+        for (int i = 1; i < array.length; i++) {
+            if (!Objects.equals(array[0], array[i])) {
```

Comment:
Avoid boxing, make use of the fact it is a non-null primitive:
```suggestion
            if (array[0] != array[i]) {
```